src/unix/syscall.c

```diff
@@ -451,8 +451,12 @@ static void file_io_complete(file f, range r, boolean is_file_offset, sg_list sg
         u64 len = range_span(r);
         len -= sg_total_len(sg);
         u64 file_len = fsfile_get_length(f->fsf);
-        if (r.start + len > file_len)   /* can happen with direct I/O */
+        if ((len < 0) || (r.start >= file_len))
```

Comment:
`len` can never be less than 0, so this should be simply `if (r.start >= file_len)`